src/constraints/ObjectConstrains.ts

```diff
@@ -0,0 +1,53 @@
+import get from 'lodash/get';
+import { ExpectedConstraintError } from '../lib/errors/ExpectedConstraintError';
+import { Result } from '../lib/Result';
+import type { BaseValidator } from '../type-exports';
+import type { IConstraint } from './type-exports';
+
+export type ObjectConstraintName = `s.object(T.when)`;
+
+export type WhenKey = PropertyKey | PropertyKey[];
+
+export interface WhenOptions<T extends BaseValidator<any>, Key extends WhenKey> {
+	is?: boolean | ((value: Key extends Array<any> ? any[] : any) => boolean);
+	then: (predicate: T) => T;
+	otherwise?: (predicate: T) => T;
```

Comment:
Isn't a `when()` with just `then` do the same as a `default()`? If yes, could we make a type that somehow says "either is or otherwise must be present when using when"?

Reply:
I don't think so? Maybe a practical example with some work code that uses Yup helps:
```ts
  depatureDate: date()
    .when([openAccordionsDepartureArrivalDateTimeKey, 'departureOrArrivalToggle'], {
      is: (openAccordionsDepartureArrivalDateTimeValue: boolean, departureOrArrivalToggleValue: DepartureOrArrivalToggleUnion) =>
        openAccordionsDepartureArrivalDateTimeValue && departureOrArrivalToggleValue === 'departure',
      then: (schema) =>
        schema
          .min(subDays(startOfDay(new Date()), 1), 'Departure date may not be more than 1 day ago.')
          .max(startOfDay(new Date('2100-01-01T00:00:00Z')), 'Departure date may not be after the year 2099.')
          .defined('Departure date is required.')
          .transform(invalidDateToNull)
          .test(
            'departure-date-is-required-when-departure-time-is-provided',
            'Departure date is required when departure time has been provided.',
            (value, context) => testDepartureArrivalDateTimeInvariableRequired(value, context, 'departureTime')
          )
    })
    .nullable(),
```

Or for another form:
```ts
  arrivalDate: date()
    .when(['departureOrArrival', 'trainIdentifications'], {
      is: (departureOrArrivalValue: Capitalize<DepartureOrArrivalToggleUnion>, trainIdentifications: SchemaRequestNewTrainIdentifications) =>
        departureOrArrivalValue === 'Arrival' && !isMultipleDaysRequest(trainIdentifications),
      then: (schema) =>
        schema
          .test('is-not-before-or-after-train-date-minus-1-day', 'Arrival date should be one day before or after the train date.', testVertrekAankomstDatumTijd)
          .transform(invalidDateToNull)
          .required('Arrival date is required.')
    })
    .nullable()
    .defined(),
```